### ISSUE TITLE
EqualsHashCode is overly strict

### ISSUE THREAD

**Issue by peterdemaeyer** [checkstyle/checkstyle#6127]
`EqualsHashCode` is overly strict. It fails not only when there is an `equals` without `hashCode`, but also when there is a `hashCode` without `equals`. The former is a bug, the latter not necessarily. This behavior has changed somewhere in between Checkstyle 6.18 and 8.12.

Expected: a way to configure Checkstyle to fail on `equals` without `hashCode`, but _not_ on `hashCode` without `equals`. A potential solution could be to extend `EqualsHashCode` with an additional property.

Workaround: add `equals` methods on all affected classes and delegate the call to `super.equals`. That makes Checkstyle happy again, but introduces useless code.

Template for bug reporting:

```bash
peter@pangolin:~/IdeaProjects/Checkstyle#6127$ javac src/HashCodeNoEquals.java 
peter@pangolin:~/IdeaProjects/Checkstyle#6127$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="EqualsHashCode"/>
    </module>
</module>
peter@pangolin:~/IdeaProjects/Checkstyle#6127$ cat src/HashCodeNoEquals.java 
public class HashCodeNoEquals {

    @Override
    public int hashCode() {
        return super.hashCode();
    }
}

$ java $RUN_LOCALE -jar checkstyle-8.13-all.jar -c config.xml src/HashCodeNoEquals.java 
Starting audit...
[ERROR] /home/.../HashCodeNoEquals.java:6:5: Definition of 'hashCode()' without corresponding definition of 'equals()'. [EqualsHashCode]
Audit done.
Checkstyle ends with 1 errors.
```

Expected: an option to configure the check to report a violation for `hashCode` without equals but not for equals without `hashCode`.

Actual: no such option, the workaround is to use suppressions as suggested by @rnveach (I didn't test it) or just implement `hashCode` everywhere with a silly `super.hashCode`.

**Comment by rnveach** [comment]
@peterdemaeyer You can use suppressions to just suppress the violations you want to ignore. It is a unique message.

**Comment by romani** [comment]
@peterdemaeyer , please report issue as requested https://github.com/checkstyle/checkstyle/blob/master/.github/ISSUE_TEMPLATE.md

if issue is not reported as required ... we close it, as it is unlikely smb will look at it.

**Comment by romani** [comment]
@peterdemaeyer, please prove that your expectations are valid, as far as I remember if any presents, both required to satisfy contract for comparison in collections.

Looks like we need to put some links to http://checkstyle.sourceforge.net/config_coding.html#EqualsHashCode to make it clear what is contract.

**Comment by peterdemaeyer** [comment]
See Josh Bloch's "Effective Java". You can find an explanation also here: https://stackoverflow.com/questions/2265503/why-do-i-need-to-override-the-equals-and-hashcode-methods-in-java.

The contract is: equal objects _must_ have the same hash code, but not vice versa. A uniform spreading of hash codes is desirable for optimal performance, it is not a strict functional requirement.

As a result: when you override `equals()` and not `hashCode()`, your objects can get lost in hash maps etc. Overriding `hashCode()` to return for example a constant, is not a bug and is functionally valid, although not a good idea in terms of performance. O(1) hash lookups will degrade to O(n), but still, functionally valid and not necessarily a bug.

**Comment by romani** [comment]
this was introduced in scope of #3308 , there is reference to some style guide that is pretty old. And most likely is focused on general guidance to avoid basic mistakes. But here might be cases where it is reasonable.

from Effective Java , 3 edition , "Item 11: Always override hashCode when you override equals"
> In summary, you must override hashCode every time you override equals, or
your program will not run correctly. Your hashCode method must obey the general
contract specified in Object and must do a reasonable job assigning unequal hash
codes to unequal instances.

> This one, for example, is always legal but should never be used:
// The worst possible legal hashCode implementation - never use!
`@Override` public int hashCode() { return 42; }

------

@peterdemaeyer ,
>  functionally valid and not necessarily a bug.

Please share experience on where it is useful? (please share code, or general ideas or it, or ....)
why you wan it to be applicable to all sources in your project?
Why suppression is not a workaround for you ? Do you have a lot cases like this ?

If some functionality/properties might appear in Check we need clearly explain for what cases it be used. 
FYI: Checkstyle is not always searching for bugs, it reports any code that might be bug or unclear or not-good-general-practice. Wantfixes always happens, life is not ideal and some case a very reasonable even they are braking some rules. In such cases suppressions are required to be used.
But is better to pus suppression on case where your are sure, rather than allow mistakes in other code, where non very attentive and experienced engineers are working.

**Comment by peterdemaeyer** [comment]
The examples I have are all in test code (not my own code by the way, I'm just the one introducing Checkstyle on a legacy codebase):

```
@Override public int hashCode() { return 0: }
```

As stated in your quote, it's always legal (but admittedly not a good idea in production code). I am aware that this was introduced with the fix for #3308, which in my opinion is an improvement rather than a bug. The reason why you should always override `hashCode` when there is an `equals` is very different from the reason why you should override `equals` when there is a `hashCode`. That's why I feel they should be independently configurable, but I won't insist. You may close this as "won't fix".

Suppression _is_ a valid workaround for me, I just don't like suppressions and try to avoid them when possible. In the meantime I just fixed all the test code implementations so I don't have to suppress anything in the end.

**Comment by romani** [comment]
Thanks a lot for effort to fix code.
I am closing this issue as we didn't find good reason to fix Check. If somebody find another case where is good to override only `hasCode` , please share here we might reconsider our decision.
